filebeat/registrar/registrar.go

```diff
@@ -176,6 +168,32 @@ func (r *Registrar) loadAndConvertOldState(f *os.File) bool {
 	return true
 }
 
+func convertOldStates(oldStates map[string]file.State) []file.State {
```

Comment:
The logic you have looks correct. What do you think about using a map with struct key to do the deduplication?

```
	states := map[file.StateOS]file.State{}
	for _, state := range oldStates {
		// If there is a duplicate, use the one with the larger offset.
		existingState, found := states[state.FileStateOS]
		if found && existingState.Offset > state.Offset {
			continue
		}

		// Makes timestamp the time of migration, as this is the best guess.
		state.Timestamp = time.Now()
		states[state.FileStateOS] = state
	}
```